### PR TITLE
Update: StatefulSet Replica scaling to include Patch Scale +1 endpoint

### LINKED ISSUE #98125: Write Updated test for StatefulSet Replica scaling to include Patch Scale +1 endpoint
# Progress <code>[3/6]</code>

-   [X] APISnoop org-flow : [AppStatefulsetsMocktest](https://github.com/apisnoop/ticket-writing/edit/AppsEndpoints/AppStatefulsetsMocktest.md)
-   [X] test approval issue : [kubernetes/kubernetes#](https://github.com/kubernetes/kubernetes/issues/98125)
-   [X] test pr : kuberenetes/kubernetes [#98126](https://github.com/kubernetes/kubernetes/pull/98126)
-   [ ] two weeks soak start date : testgrid-link
-   [ ] two weeks soak end date :
-   [ ] test promotion pr : kubernetes/kubernetes#?

# Identifying an untested feature Using APISnoop

According to this APIsnoop query, there are still some remaining StatefulsetScale endpoints that are untested.

with this query, you can filter untested endpoints by their category and eligibility for conformance. e.g below shows a query to find all conformance eligible untested, stable, apps endpoints

```sql-mode
SELECT
  endpoint,
  -- k8s_action,
  -- path,
  -- description,
  kind
  FROM testing.untested_stable_endpoint
  where eligible is true
    and endpoint like '%StatefulSetScale%'
  --and category = 'core'
  order by kind, endpoint desc
  limit 25;
```

```example
               endpoint                | kind
---------------------------------------|-------
 patchAppsV1NamespacedStatefulSetScale | Scale
(1 row)


# Verifying increase in coverage with APISnoop


```sql-mode
select distinct  endpoint, right(useragent,73) AS useragent
from testing.audit_event
-- where useragent ilike '%subresource%'
 where endpoint ilike '%AppsV1NamespacedStatefulSetScale%'
 and release_date::BIGINT > round(((EXTRACT(EPOCH FROM NOW()))::numeric)*1000,0) - 60000
and useragent like 'e2e%'
order by endpoint
limit 30;

```

```example
                endpoint                 |                                 useragent
-----------------------------------------|---------------------------------------------------------------------------
 patchAppsV1NamespacedStatefulSetScale   |  [StatefulSetBasic] should have a working scale subresource [Conformance]
 readAppsV1NamespacedStatefulSetScale    |  [StatefulSetBasic] should have a working scale subresource [Conformance]
 replaceAppsV1NamespacedStatefulSetScale |  [StatefulSetBasic] should have a working scale subresource [Conformance]
(3 rows)

```

# Final notes

If a test with these calls gets merged, ****test coverage will go up by 1 point****

This test is also created with the goal of conformance promotion.

---

/sig testing

/sig architecture

/area conformance

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- patchAppsV1NamespacedStatefulSetScale

**Which issue(s) this PR fixes:**
Fixes #98125

**Testgrid Link:**

**Special notes for your reviewer:**
Adds +1 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

**Comment by Riaankl** [comment]
`           	Test:       	TestAdmit/skip_webhook_whose_objectSelector_does_not_match_CRD's_labels
            	Messages:   	skip webhook whose objectSelector does not match CRD's labels: annotations not set as expected.
W0117 23:57:43.395337  127900 dispatcher.go:170] Failed calling webhook, failing open cache2: failed calling webhook "cache2": an error on the server ("webhook internal server error") has prevented the request from succeeding
E0117 23:57:43.395507  127900 dispatcher.go:171] failed calling webhook "cache2": an error on the server ("webhook internal server error") has prevented the request from succeeding
FAIL`

/test all

**Comment by Riaankl** [comment]
/test pull-kubernetes-bazel-test

`AIL: //test/conformance:conformance_test (run 2 of 2) (see /bazel-scratch/.cache/bazel/_bazel_root/7989b31489f31aee54f32688da2f0120/execroot/io_k8s_kubernetes/bazel-out/k8-fastbuild/testlogs/test/conformance/conformance_test/run_2_of_2/test.log)
FAILED: //test/conformance:conformance_test (Summary)
      /bazel-scratch/.cache/bazel/_bazel_root/7989b31489f31aee54f32688da2f0120/execroot/io_k8s_kubernetes/bazel-out/k8-fastbuild/testlogs/test/conformance/conformance_test/run_1_of_2/test.log
      /bazel-scratch/.cache/bazel/_bazel_root/7989b31489f31aee54f32688da2f0120/execroot/io_k8s_kubernetes/bazel-out/k8-fastbuild/testlogs/test/conformance/conformance_test/run_2_of_2/test.log
INFO: From Testing //test/conformance:conformance_test (run 2 of 2):`

**Comment by Riaankl** [comment]
/assign @smarterclayton
Updated the test to add `patchAppsV1NamespacedStatefulSetScale` endpoint to conformance.
Ran `/test all` several times and the update seems good to go. 
The only issue was related to my incorrect updating of the conformance.yaml now fix by @heyste 
The test metadata is updated as required by the [rules](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-comment-metadata) noting changes to conformance test in subsequent releases and now include 1.21

File: test/e2e/apps/statefulset.go
Comment (krmayankk) [inline test/e2e/apps/statefulset.go:47]: can you keep the separation, go fmt should do it

File: test/e2e/apps/statefulset.go
Comment (krmayankk) [inline test/e2e/apps/statefulset.go:888]: should we similar to the original status, also add ?
e2estatefulset.WaitForRunningAndReady(c, *ss.Spec.Replicas, ss)
waitForStatus(c, ss)

**Review by krmayankk — COMMENTED** [review]
looks good , two small suggestions

File: test/e2e/apps/statefulset.go
Comment (Riaankl) [reply test/e2e/apps/statefulset.go]: Changes as requested. Thank you for the review.

File: test/e2e/apps/statefulset.go
Comment (Riaankl) [reply test/e2e/apps/statefulset.go:890]: It is included on line 847-848. The pattern looks to be using this only at the "create" stage of the statefulset.
Does this address your concern?

**Comment by Riaankl** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd

`E0119 21:49:33.757] FAIL: pull-kubernetes-e2e-gce-ubuntu-containerd
I0119 21:49:33.757] Call:  gcloud auth activate-service-account --key-file=/etc/service-account/service-account.json
W0119 21:49:34.291] Activated service account credentials for: [prow-build@k8s-infra-prow-build.iam.gserviceaccount.com]
I0119 21:49:34.376] process 24894 exited with code 0 after 0.0m
I0119 21:49:34.376] Call:  gcloud config get-value account
I0119 21:49:34.846] process 24907 exited with code 0 after 0.0m
I0119 21:49:34.846] Will upload results to gs://kubernetes-jenkins/pr-logs using prow-build@k8s-infra-prow-build.iam.gserviceaccount.com
I0119 21:49:34.846] Upload result and artifacts...
I0119 21:49:34.846] Gubernator results at https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/98126/pull-kubernetes-e2e-gce-ubuntu-containerd/1351635935267131392
I0119 21:49:34.847] Call:  gsutil ls gs://kubernetes-jenkins/pr-logs/pull/98126/pull-kubernetes-e2e-gce-ubuntu-containerd/1351635935267131392/artifacts
W0119 21:49:35.681] CommandException: One or more URLs matched no objects.`

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
6 flakes
```
Kubernetes  e2e suite: [sig-storage] CSI mock volume CSI FSGroupPolicy [LinuxOnly]  should not modify fsGroup if fsGroupPolicy=None
Kubernetes  e2e suite: [sig-api-machinery] CustomResourcePublishOpenAPI  [Privileged:ClusterAdmin] works for multiple CRDs of different groups  [Conformance]
Kubernetes  e2e suite: [sig-api-machinery] AdmissionWebhook  [Privileged:ClusterAdmin] should be able to deny attaching pod  [Conformance]
Kubernetes  e2e suite: [sig-cli] Kubectl client Update Demo should create and stop a replication controller  [Conformance]
Kubernetes  e2e suite: [sig-storage] In-tree Volumes [Driver:  local][LocalVolumeType: block] [Testpattern: Pre-provisioned PV (default  fs)] subPath should support existing single file [LinuxOnly]
Kubernetes  e2e suite: [sig-storage] Subpath Container restart should verify that  container can restart successfully after configmaps modified
```

common error message
```
dial tcp 34.83.220.175:443: connect: connection timed out
```

**Comment by Riaankl** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
3 flakes not related to this PR

`Kubernetes e2e suite: [sig-storage] Volumes ConfigMap should be mountable expand_more	2m38s
Kubernetes e2e suite: [sig-storage] PersistentVolumes-local [Volume type: dir-bindmounted] Two pods mounting a local volume at the same time should be able to write from pod1 and read from pod2 expand_more	2m45s
Kubernetes e2e suite: [sig-cli] Kubectl client Kubectl diff should check if kubectl diff finds a difference for Deployments [Conformance] expand_more'

`e2e.go: Test expand_less | 21m35s
-- | --
error during ./hack/ginkgo-e2e.sh --ginkgo.skip=\[Slow\]\|\[Serial\]\|\[Disruptive\]\|\[Flaky\]\|\[Feature:.+\] --minStartupPods=8 --report-dir=/workspace/_artifacts --disable-log-dump=true: exit status 1`

File: test/e2e/apps/statefulset.go
Comment (krmayankk) [reply test/e2e/apps/statefulset.go:890]: agree its not the pattern, but extra validation makes the test more robust, that indeed after patching the scale subresource, the StatefulSet pod is ready and good to go

Comment (Riaankl) [reply test/e2e/apps/statefulset.go:890]: Thank you for helping to make this test more solid.
I believe `WaitForRunningAndReady` is for creating new resources.
However, I think adding ` WaitForStatusReplicas` will add the value you would like to see.
I update with ` WaitForStatusReplicas` and I trust this will make the test ready to be merged.

**Comment by Riaankl** [comment]
`test pull-kubernetes-conformance-image-test` is a known issue and will be fixed by WIP in #94015. This should not be a blocker for this PR. It is also not a `Required` job
https://github.com/kubernetes/kubernetes/pull/95503#issuecomment-762075575

**Comment by Riaankl** [comment]
> looks good , two small suggestions

@krmayankk we made the proposed updates. Thank you for the input.
Is this PR good to be merged?

**Comment by krmayankk** [comment]
/lgtm
@kow3ns @soltysh @tnozicka for approval

File: test/e2e/apps/statefulset.go
Comment (soltysh) [inline test/e2e/apps/statefulset.go:856]: Unnecessary change, passing empty string is equal not to pass anything.

File: test/e2e/apps/statefulset.go
Comment (soltysh) [inline test/e2e/apps/statefulset.go:865]: Ditto.

File: test/e2e/apps/statefulset.go
Comment (soltysh) [inline test/e2e/apps/statefulset.go:872]: Ditto.

File: test/e2e/apps/statefulset.go
Comment (soltysh) [inline test/e2e/apps/statefulset.go:891]: I don't think we need to wait for the replicas here, the other part of this test only ensures that the update worked. I think we can safely say that ensuring the patch operation worked is sufficient as well. You don't rely nor wait for results from the rollout. I'd drop this line, this will safe unnecessary wait time during this tests. There are other tests ensuring that scaling actually works.

**Review by soltysh — COMMENTED** [review]
/lgtm cancel
a couple of nits, also pls squash your changes into a single commit

File: test/e2e/apps/statefulset.go
Comment (soltysh) [reply test/e2e/apps/statefulset.go:890]: > agree its not the pattern, but extra validation makes the test more robust, that indeed after patching the scale subresource, the StatefulSet pod is ready and good to go

@krmayankk iirc there are other tests that ensure the scale is working as expected. The purpose of this test is to just ensure the update & patch API calls are working as expected. That additional scrutiny imo is unnecessary b/c we don't rely on the result at all and that wail will only make this test last longer. So I suggested dropping that additional wait.

File: test/e2e/apps/statefulset.go
Comment (Riaankl) [reply test/e2e/apps/statefulset.go:865]: The extra `"  " ` was an automatic `gofmt` change that I can remove.

File: test/e2e/apps/statefulset.go
Comment (Riaankl) [reply test/e2e/apps/statefulset.go:872]: Yes, same here

File: test/e2e/apps/statefulset.go
Comment (Riaankl) [reply test/e2e/apps/statefulset.go:856]: @krmayankk advised we should add the check to guard against test flakiness. Which is our worst enemy in running these tests. https://github.com/kubernetes/kubernetes/pull/98126#discussion_r559879297
Should we keep it?

**Comment by Riaankl** [comment]
@soltysh Updated and squashed the commits.
Thank you for the review. Should be good to go after the jobs ran through.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Comment by Riaankl** [comment]
/test pull-kubernetes-e2e-kind
Unrelated flake
```
[Fail] [sig-cli] Kubectl client Simple pod [It] should return command exit codes 
test/e2e/kubectl/kubectl.go:515
```

**Comment by Riaankl** [comment]
/test pull-kubernetes-node-e2e

Unrelated flake
```
[Fail] [k8s.io] Summary API [NodeConformance] when querying /stats/summary [It] should report resource usage through the stats api 
I0211 17:04:00.746] _output/local/go/src/k8s.io/kubernetes/test/e2e_node/summary_test.go:344
```

**Comment by dims** [comment]
looks like you need to run hack/update-gofmt.sh @Riaankl

**Comment by dims** [comment]
/lgtm
/approve

/assign @smarterclayton @spiffxp 

Aaron, Clayton, this is ready to merge

**Comment by Riaankl** [comment]
Sorry for the confusion. `pull-kubernetes-verify` failed on a `gofmt` error.
Fixed and pushed

**Review by dims — APPROVED** [review]
/lgtm

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

File: test/e2e/apps/statefulset.go
Comment (spiffxp) [reply test/e2e/apps/statefulset.go:890]: /hold
where's the e2e/conformance test that verifies stateful sets scale as expected?  as an end user I'd be surprised if the API allowed me to update the spec, but the status never reconciled

**Comment by spiffxp** [comment]
I put on a /hold because I'd like to verify whether other conformance tests that update the scale subresource wait for status to reconcile.  I'm not sure "read after write works" is sufficient to say stateful sets behave as expected

**Comment by spiffxp** [comment]
/hold cancel
There are two statefulset conformance tests that exercise scaling by updating `spec.Replicas` directly
- `Scaling should happen in predictable order and halt if any stateful pod is unhealthy [Slow]`
- `Burst scaling should run to completion even with unhealthy pods [Slow]`

So agree with @soltysh, it's fine if this verifies the behavior that `spec.Replicas` gets updated, since those tests cover the behavior of scaling actually happening.

**Pull request merged**
